### PR TITLE
Onboard Mixtral to bidirectional conversion

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Onboard Mixtral to bidirectional conversion, b/452391836.
* Add mappings, include `MIXTRAL_HF_WEIGHTS_TO_SHAPE`, `MIXTRAL_MAXTEXT_TO_HF_PARAM_MAPPING`, and `MIXTRAL_MAXTEXT_TO_HF_PARAM_HOOK_FN`.
* Other small fixes

# Tests

Mainly tested smaller model 8x7B:

* Unscan
  * **HF --> MaxText** - threshold aligned with [script](https://github.com/AI-Hypercomputer/maxtext/blob/56a7fd8f7618901edc93121a5e9bb3a9919ee623/end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh#L42) with extra `max_kl_div=0.015`: [test](https://paste.googleplex.com/4828500219527168)
  * **MaxText --> HF** - use above checkpoint and converted back with `rtol=1e-2, atol=1e-2`: [test](https://paste.googleplex.com/5684908511068160)
* Scan
  * **HF --> MaxText** - load checkpoint in SFT: [test](https://screenshot.googleplex.com/6etbzHQxeRoRfKn), logits verification with extra `max_kl_div=0.015` [test](https://paste.googleplex.com/5709549396951040)
  * **MaxText --> HF** - use above checkpoint and converted back with `rtol=1e-2, atol=1e-2`: [test](https://paste.googleplex.com/6530292762411008)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/utils/ckpt_conversion/README.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ The following models are supported:
 - Gemma2 (2B, 9B, 27B).
 - Gemma3 multimodal (4B, 12B, 27B).
 - Qwen3 (0.6B, 4B, 8B, 14B, 32B).
+- Mixtral (8x7B, 8X22B).
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/README.md]: nit: 8x22B

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py [hunk 2]
```diff
@@ -1424,6 +1424,165 @@ def transform_query_kernel(arr):
...
+
+  def scale_query_layer(input_tensor, target_shape):
+    if saving_to_hf:
+      depth_scale = np.dtype("float32").type(np.sqrt(maxtext_config.head_dim))
+      return (input_tensor * depth_scale).astype(input_tensor.dtype)
+    else:
+      depth_scale = np.dtype("float32").type(1 / np.sqrt(maxtext_config.head_dim))
+      return (input_tensor * depth_scale).astype(input_tensor.dtype)
+
+  plan = [
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: I would prefer that we separate the scan and unscan cases, but up to you

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py [hunk 3]
```diff
@@ -1424,6 +1424,165 @@ def transform_query_kernel(arr):
...
+    def chained_hook(x, shape):
+      for fn in funcs:
+        x = fn(x, shape)
+      return x
+
+    return chained_hook
+
+  for maxtext_pattern, op_val in plan:
+    op_funcs = op_val if isinstance(op_val, list) else [op_val]
+    hook_fn = create_chained_hook(op_funcs)
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: Most likely, you can directly return a list of hooks (without `create_chained_hook`), similar to gemma2:
https://github.com/AI-Hypercomputer/maxtext/blob/948d302a36f73305f8bb6b7c37eb4c2e86f070ea/src/MaxText/utils/ckpt_conversion/utils/param_mapping.py#L534-L537

The logic is handled by `apply_hook_fns`, which is used by `to_maxtext` and `to_huggingface` (via _process)
https://github.com/AI-Hypercomputer/maxtext/blob/948d302a36f73305f8bb6b7c37eb4c2e86f070ea/src/MaxText/utils/ckpt_conversion/utils/utils.py#L248-L255

File: src/MaxText/utils/ckpt_conversion/utils/utils.py [hunk 4]
```diff
@@ -195,7 +197,8 @@ def process_maxtext_param(
 
   # Case 3 or 4: The source tensor is stacked on a single axis.
   # We determine if it's an unscanned MoE (expert axis) or standard scanned (layer axis).
-  is_unscanned_moe = "moe_block" in maxtext_param_key and any(
+  moe_block_list = ["moe_block", "MoeBlock_0-w"]
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/utils/utils.py:202]: Curious why it is `MoeBlock_0-w` rather than `MoeBlock_0`

File: tests/forward_pass_logit_checker.py [hunk 5]
```diff
@@ -380,7 +380,7 @@ def main(config, test_args):  # pylint: disable=W0621
       raise ValueError("run_hf_model requires hf_model_path")
     hf_model = AutoModelForCausalLM.from_pretrained(test_args.hf_model_path, dtype=torch.bfloat16)
     tokenizer = AutoTokenizer.from_pretrained(test_args.hf_model_path)
-    if "Llama-3.1" in test_args.hf_model_path:
+    if tokenizer.pad_token is None or "Llama-3.1" in test_args.hf_model_path:
```
Comment (shuningjin) [inline tests/forward_pass_logit_checker.py]: Could you add a comment to say mixtral has none pad_token?

**Review by shuningjin — COMMENTED** [review]
Thank you for onboarding this! Could you also run forward logit check for "Scan: HF -> MaxText"?

File: src/MaxText/utils/ckpt_conversion/README.md (hunk 1 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/README.md]: Nice catch!

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: No problem, I will branch them out.

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: I recall I met issues earlier, but I will have another try.

File: src/MaxText/utils/ckpt_conversion/utils/utils.py (hunk 4 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:202]: Yes, `w` is needed for weights, and except for gate. Gate values are stack in layers only, but weights are stack in both expert and layer. 

I suspect something is wrong in Qwen3-moe conversion part. It seems to me the step 3 logit verification is re-use the HF model instead of generated safetensor checkpoint in https://github.com/AI-Hypercomputer/maxtext/pull/2177.

File: tests/forward_pass_logit_checker.py (hunk 5 above)
Comment (RissyRan) [reply tests/forward_pass_logit_checker.py]: I will update it something like bellow:

```
    pad_token_models = ["Llama-3.1", "Mixtral-8x"]
    if any(model in test_args.hf_model_path for model in pad_token_models):
      tokenizer.pad_token = tokenizer.eos_token
```

File: src/MaxText/utils/ckpt_conversion/utils/param_mapping.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/param_mapping.py]: Thanks for pointing out! I have another try, and forward logits tests also match after removal. https://paste.googleplex.com/6122812572041216

**Comment by RissyRan** [comment]
> Thank you for onboarding this! Could you also run forward logit check for "Scan: HF -> MaxText"?

Yes, it matches as well. Thanks for pointing it out! https://paste.googleplex.com/5709549396951040

File: src/MaxText/utils/ckpt_conversion/utils/utils.py (hunk 4 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:202]: Could you add a comment here (`w is needed for weights, and except for gate. Gate values are stack in layers only, but weights are stack in both expert and layer`) for future reference?

Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/utils/utils.py:202]: cc @parambole and @Rohan-Bierneni for your reference in conversion verification.
